feat(ci): modify GetCIStatus to utilize GitHub GraphQL API

dblinkhorn opened the pull request:
This API allows us to get both checks and statuses in one call, and gives us an overall status of all the CI checks. This means we can get rid of the `getCIChecksResult` and `getCombinedStatus` functions. Additionally, `GetCIStatus` will now individually log any failed checks or statuses from the query.

Here are some results from testing:

- Query on commit with statuses & checks
```
DEBU[2025-01-14 16:34:32] Debug logging enabled                        
DEBU[2025-01-14 16:34:32] Will use GitHub token for authentication     
DEBU[2025-01-14 16:34:32] Using token starting with ghp_8kXW80...      
DEBU[2025-01-14 16:34:32] Using GitHub token for authentication        
INFO[2025-01-14 16:34:32] Checking CI status on dblinkhorn/wait-for-github-graphql-test@23c27195d5172ddf9a4bef27243370e428548189 
2025/01/14 16:34:32 [DEBUG] POST https://api.github.com/graphql
CI successful
```

- Querying a repo without statuses or checks
```
DEBU[2025-01-14 16:25:43] Debug logging enabled                        
DEBU[2025-01-14 16:25:43] Will use GitHub token for authentication     
DEBU[2025-01-14 16:25:43] Using token starting with ghp_8kXW80...      
DEBU[2025-01-14 16:25:43] Using GitHub token for authentication        
INFO[2025-01-14 16:25:43] Checking CI status on dblinkhorn/wait-for-github-graphql-test@8861c74e96251fac19cf94305c9ca0475b451d0c 
2025/01/14 16:25:43 [DEBUG] POST https://api.github.com/graphql
INFO[2025-01-14 16:25:43] Did not find any checks and/or statuses. Retrying in 5s to see if any appear 
2025/01/14 16:25:48 [DEBUG] POST https://api.github.com/graphql
DEBU[2025-01-14 16:25:48] No checks or statuses found after retry, assuming success 
CI successful
```

- Querying commit with a failed check & status
```
DEBU[2025-01-14 16:04:43] Debug logging enabled                        
DEBU[2025-01-14 16:04:43] Will use GitHub token for authentication     
DEBU[2025-01-14 16:04:43] Using token starting with ghp_8kXW80...      
DEBU[2025-01-14 16:04:43] Using GitHub token for authentication        
INFO[2025-01-14 16:04:43] Checking CI status on dblinkhorn/wait-for-github-graphql-test@63dd76c531777f14d2614cf3642b3316004d8ffb 
2025/01/14 16:04:43 [DEBUG] POST https://api.github.com/graphql
DEBU[2025-01-14 16:04:43] CI failed. Failed nodes:                     
DEBU[2025-01-14 16:04:43] CheckRun 'build' failed                      
DEBU[2025-01-14 16:04:43] StatusContext 'security-check' failed        
CI failed. Please check CI on the following commit: https://github.com/dblinkhorn/wait-for-github-graphql-test/commit/63dd76c531777f14d2614cf3642b3316004d8ffb
```

- Query commit with skipped check
```
DEBU[2025-01-14 16:39:01] Debug logging enabled                        
DEBU[2025-01-14 16:39:01] Will use GitHub token for authentication     
DEBU[2025-01-14 16:39:01] Using token starting with ghp_8kXW80...      
DEBU[2025-01-14 16:39:01] Using GitHub token for authentication        
INFO[2025-01-14 16:39:01] Checking CI status on dblinkhorn/wait-for-github-graphql-test@b6f92ebb708854dbb9cd5ce1333e081a9e9cc943 
2025/01/14 16:39:01 [DEBUG] POST https://api.github.com/graphql
CI successful
```

- Immediately query for CI status after a new commit is pushed to repo
```
$ git push & ~/wait-for-github/wait-for-github --github-token ghp_8***y --log-level debug ci https://github.com/dblinkhorn/wait-for-github-graphql-test/commit/$(git rev-parse HEAD)                    
[1] 6883
DEBU[2025-01-14 16:43:41] Debug logging enabled                        
DEBU[2025-01-14 16:43:41] Will use GitHub token for authentication     
DEBU[2025-01-14 16:43:41] Using token starting with ghp_8kXW80...      
DEBU[2025-01-14 16:43:41] Using GitHub token for authentication        
INFO[2025-01-14 16:43:41] Checking CI status on dblinkhorn/wait-for-github-graphql-test@1f75ceb24f804cebdfdda2e72b7c54a7169c0bc5 
2025/01/14 16:43:41 [DEBUG] POST https://api.github.com/graphql
INFO[2025-01-14 16:43:41] Did not find any checks and/or statuses. Retrying in 5s to see if any appear 
Enumerating objects: 1, done.
Counting objects: 100% (1/1), done.
Writing objects: 100% (1/1), 177 bytes | 177.00 KiB/s, done.
Total 1 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/dblinkhorn/wait-for-github-graphql-test.git
   b6f92eb..1f75ceb  main -> main
[1]  + 6883 done       git push
2025/01/14 16:43:46 [DEBUG] POST https://api.github.com/graphql
INFO[2025-01-14 16:43:47] CI is not finished yet, rechecking in 30s    
2025/01/14 16:44:11 [DEBUG] POST https://api.github.com/graphql
CI successful
```

iainlane commented on internal/github/github.go:
```diff
@@ -164,138 +188,114 @@ func (c CIStatus) String() string {
+		hasChecks = false
+		hasStatuses = false
+		failedNodes = []string{}
+
```
(I've not reviewed the rest of this at all! Anyone else can feel free to. I just saw that you got a CI failure.)

You are creating this variable on line 231 above and then resetting it here. The value from the intiial creation is never used - that's why the linter is complaining. I think 231 can be dropped and this replaced with a `:=`.

dblinkhorn replied:
Yes just realized this as well. I just moved the declaration inside the for loop since I was resetting every loop anyway and only accessing within the scope of the loop. Thank you!

iainlane commented on internal/github/github.go:
```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
-				return CIStatusPassed, nil
-			}
+		// return early if all checks and statuses are successful. no need to evaluate individual nodes in the response.
+		if strings.ToLower(query.Repository.Object.Commit.StatusCheckRollup.State) == "success" {
```
Suggest extracting `query.Repository.Object.Commit.StatusCheckRollup` to a variable as you access it a few times & this would make the code a bit easier to read.

zerok commented on internal/github/github.go:
```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
-				return CIStatusPassed, nil
-			}
+		// return early if all checks and statuses are successful. no need to evaluate individual nodes in the response.
+		if strings.ToLower(query.Repository.Object.Commit.StatusCheckRollup.State) == "success" {
```
Please give `query.Repository.Object.Commit.StatusCheckRollup` its own variable. This should make the checks much easier to read 🙂

dblinkhorn replied:
Fixed. Thanks!

zerok commented on internal/github/github.go:
```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
+			switch node.Typename {
+			case "CheckRun":
+				hasChecks = true
+				if strings.ToLower(node.CheckRun.Conclusion) == "failure" {
```
Please create constants for the possible conclusions and statuses 🙂

dblinkhorn replied:
Done.

iainlane commented on internal/github/github_test.go:
```diff
@@ -297,201 +304,307 @@ func TestIsPRMergedOrClosed_Error(t *testing.T) {
 
-			ctx := context.Background()
+			mockServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if r.Method == "POST" && r.URL.Path == "/graphql" {
+					w.Header().Set("Content-Type", "application/json")
+					_, _ = w.Write([]byte(tt.mockGraphQL))
+				} else {
+					http.NotFound(w, r)
+				}
+			}))
```
this is a nit/styleish point. In my experience a lot of Go developers write code so that the "normal" path is outside of any `if` blocks. A good rule of thumb is to try to keep it as left aligned as possible, and it also usually means you don't need an `else` part to your condition. Here it would be like this (sorry for any typos, I just did this in the editor here rather than trying it):

```suggestion
				if r.Method != "POST" || r.URL.Path != "/graphql" {
					http.NotFound(w, r)
					return
				}

				w.Header().Set("Content-Type", "application/json")
				_, _ = w.Write([]byte(tt.mockGraphQL))
			}))
```

See what I mean? The unusual/error/less common path is indented, and the normal one is aligned to the left. It means that people reading your code get a nice visual cue as to the flow just by looking at the shape of the text. 🙂

dblinkhorn replied:
Fixed. Much easier to read, thanks.

iainlane commented on internal/github/github.go:
```diff
@@ -107,9 +116,16 @@ func AuthenticateWithApp(ctx context.Context, privateKey []byte, appID, installa
 
-	return GHClient{client: githubClient, pendingRecheckTime: pendingRecheckTime}, nil
+	restClient := github.NewClient(httpClient)
+	graphQLClient := graphql.NewClient("https://api.github.com/graphql", httpClient)
```
Could you add a test to make sure that retrying works for this client please? I don't believe caching really does for GraphQL but retrying should!

dblinkhorn replied:
Test added. Thanks for noticing we weren't testing this for the GraphQL client.

iainlane reviewed (commented):
Thanks for working on this! It looks really good. Don't worry that I gave a bunch of comments - I never met a PR that I didn't have something to say about (it's not you it's me 😁).

dblinkhorn commented:
> Thanks for working on this! It looks really good. Don't worry that I gave a bunch of comments - I never met a PR that I didn't have something to say about (it's not you it's me 😁).

Please leave as many as you like. I'm new to Go and this tool so I appreciate any and all feedback on how improvements can be made. Thanks!

dblinkhorn commented on internal/github/github.go:
```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
-			if state != "pending" {
-				return CIStatusUnknown, nil
-			}
+		if strings.ToLower(query.Repository.Object.Commit.StatusCheckRollup.State) == "pending" {
```
I was not sure about the pending status situation...it was my impression that the pending status would appear in the scenario where the CI is queried prior to checks and/or statuses initializing, which would mean we should force a retry after the `pendingRecheckTime`.

dblinkhorn commented on internal/github/github.go:
```diff
@@ -164,138 +180,113 @@ func (c CIStatus) String() string {
-	listOptions := github.ListOptions{
-		PerPage: 100,
+func (c GHClient) GetCIStatus(ctx context.Context, owner, repoName, ref string) (CIStatus, error) {
+	var query struct {
+		Repository struct {
+			Object struct {
+				Commit struct {
+					StatusCheckRollup struct {
+						State    string `graphql:"state"`
+						Contexts struct {
+							Nodes []struct {
+								Typename string `graphql:"__typename"`
+								CheckRun struct {
+									Name       string `graphql:"name"`
+									Conclusion string `graphql:"conclusion"`
+								} `graphql:"... on CheckRun"`
+								StatusContext struct {
+									Context string `graphql:"context"`
+									State   string `graphql:"state"`
+								} `graphql:"... on StatusContext"`
+							} `graphql:"nodes"`
+							PageInfo struct {
+								EndCursor   *string `graphql:"endCursor"`
+								HasNextPage bool    `graphql:"hasNextPage"`
+							} `graphql:"pageInfo"`
+						} `graphql:"contexts(first: 100, after: $cursor)"`
+					} `graphql:"statusCheckRollup"`
+				} `graphql:"... on Commit"`
+			} `graphql:"object(expression: $ref)"`
+		} `graphql:"repository(owner: $owner, name: $repository)"`
 	}
 
```
Thanks for this tip! I want to break this down into smaller parts per your suggestion. I was thinking of just defining multiple types at the top of the file which represent the sub-structs of the larger query...what do you think? We might have cause to re-use some of the sub-structs in future functionality:

```go
type PageInfo struct {
    EndCursor   *string
    HasNextPage bool
}

type CheckRun struct {
    Name
    Conclusion string
}

type StatusContext struct {
    Context string
    State   string
}

type ContextNode struct {
    Typename      string        `graphql:"__typename"`
    CheckRun      CheckRun      `graphql:"... on CheckRun"`
    StatusContext StatusContext `graphql:"... on StatusContext"`
}

type Contexts struct {
    Nodes    []ContextNode
    PageInfo PageInfo
}

type StatusCheckRollup struct {
    State    string
    Contexts Contexts `graphql:"contexts(first: 100, after: $cursor)"`
}

type Commit struct {
    StatusCheckRollup StatusCheckRollup
}

type Object struct {
    Commit Commit `graphql:"... on Commit"`
}

type Repository struct {
    Object Object `graphql:"object(expression: $ref)"`
}

// overall query
var query struct {
    Repository Repository `graphql:"repository(owner: $owner, name: $repository)"`
}
```

iainlane replied:
Yes indeed! That might be a bit too fine-grained - you might end up jumping around when reading it - but play around with a few options and see what makes sense.

What's kind of cool about the "one big query" is that it visually gives you an idea of the shape of the data. When flattening everything out that becomes a bit less obvious.

Maybe starting at `StatusCheckRollup` since that's where the objects begin to have more than one field?

Kind of waxing lyrical here - this is an exercise in readability and a matter of taste which doesn't matter too much at the end of the day, and can always be changed. 🙂

dblinkhorn replied:
Reduced the granularity of the deconstruction. Lemme know what you think.

iainlane commented on internal/github/github_test.go:
```diff
@@ -103,9 +105,9 @@ func TestNewGithubClientWithAppAuthentication(t *testing.T) {
 }
 
-// newClientFromMock returns a new GHClient whose transport is configured to use
+// newClientFromMock returns a new REST & GraphQL GHClient whose transport is configured to use
```
ok we're doing micro nits now 😈 

```suggestion
// newClientFromMock returns a new REST & GraphQL GHClient whose transports are configured to use
```

iainlane commented on internal/github/github.go:
```diff
@@ -95,9 +129,15 @@ func AuthenticateWithToken(ctx context.Context, token string, pendingRecheckTime
 
-	return GHClient{client: githubClient, pendingRecheckTime: pendingRecheckTime}
+	restClient := github.NewClient(httpClient)
+	graphQLClient := graphql.NewClient("https://api.github.com/graphql", httpClient)
```
I noticed this client isn't using the proper logger ...

```console
INFO[2025-01-23 18:17:40] Checking CI status on iainlane/test@7456abc7104e4138951a72b6ca0dc0039d58ed62 
2025/01/23 18:17:40 [DEBUG] POST https://api.github.com/graphql
CI successful
```

See the two styles of message?

We're using logrus in `wait-for-github` but that doesn't plug straight into `go-retryablehttp`'s `Logger` field. So I think we should fix this separately, maybe by switching to `slog`, which does work fine.

I think it's more appropriate to do that in a follow-up PR, as it's a separate issue.

iainlane commented on internal/github/github.go:
```diff
@@ -164,138 +211,107 @@ func (c CIStatus) String() string {
 
 	for {
-		runs, resp, err := c.client.Checks.ListCheckRunsForRef(ctx, owner, repoName, ref, opt)
+		err := c.graphQLClient.Query(ctx, &query, vars)
```
Did I say this before? I think I did but I can't find it now...

We could save rate limit quota if we first do a minimal check perhaps where we ask for the rollup only. And then, only when we need to, should we do the full one and ask for details of all the checks. Basically in some cases we can skip a lot of the work and that's good for our tokens.

A plan could be: we merge this as it is now, test the new state `main` with our internal usecases, follow up with this fix, and then release a new version once it's in. WDYT?

dblinkhorn replied:
Yes that sounds like a good idea.

iainlane reviewed (approved):
Great, good work. I think we should merge this in now.

Just one thing before you do: could you rebase your change, _squash_ all the commits together, and then write a nice detailed commit message explaining what's being changed. I've been trying to keep the history clean and readable here. 🙂

I left a few suggestions but 2/3 of them are suggestions for follow-ups and one is a micro nit. No _need_ for me to do another review, but always feel free to ask for one anyway if you want!

dblinkhorn commented:
Squashed commits and added a detailed commit message. Created [this issue](https://github.com/grafana/wait-for-github/issues/193) to address your comment on the initial rollup check, and [this issue](https://github.com/grafana/wait-for-github/issues/194) to address your comment on the logging inconsistency.
